### ISSUE TITLE
aws_eks_node_group not performing version updates?

### ISSUE THREAD

**Issue by gygitlab** [hashicorp/terraform-provider-aws#26345]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform CLI and Terraform AWS Provider Version

```
Terraform v1.1.6
on darwin_arm64
+ provider registry.terraform.io/hashicorp/aws v4.18.0
+ provider registry.terraform.io/hashicorp/tls v3.4.0
```

### Affected Resource(s)

* aws_eks_node_group

### Terraform Configuration Files

Please include all Terraform configurations required to reproduce the bug. Bug reports without a functional reproduction may be closed without investigation.

```hcl
resource "aws_eks_node_group" "web_pool" {
  cluster_name = aws_eks_cluster.cluster[0].name

  ami_type            = local.eks_ami_type
  version               = aws_eks_cluster.cluster[0].version
  release_version = var.eks_node_group_ami_release_version  # null in this case

  node_group_name_prefix = "web_pool_"
  node_role_arn                   = aws_iam_role.eks_node_role[0].arn
  subnet_ids                         = local.eks_backend_node_subnet_ids
  instance_types                  = [var.web_node_pool_instance_type]
  disk_size                           = local.eks_custom_ami ? null : var.web_node_pool_disk_size

  labels = {
    workload = "web"
  }

  tags = merge({
    "k8s.io/cluster-autoscaler/${aws_eks_cluster.cluster[0].name}" = "owned"
    "k8s.io/cluster-autoscaler/enabled"                                   = "TRUE"
  }, var.additional_tags)

  lifecycle {
    ignore_changes = [
      scaling_config[0].desired_size,
    ]
  }
}
```

### Debug Output

There's no change reported.

### Expected Behavior

According to the docs around versions it seems like `release_version` should be defaulting to the latest AMI:

>>>
- [release_version](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/eks_node_group#release_version) – (Optional) AMI version of the EKS Node Group. Defaults to latest version for Kubernetes version.
- [version](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/eks_node_group#version) – (Optional) Kubernetes version. Defaults to EKS Cluster Kubernetes version. Terraform will only perform drift detection if a configuration value is provided.
>>>

My interpretation of this was that we need to specify `version` to trigger Terraform drift detection that in turn would update `release_version`. [This previous issue looks to match this understanding](https://github.com/hashicorp/terraform-provider-aws/issues/12675).

However this doesn't look to be the case and `release_version` is not changing after initial creation. If this is incorrect is there no way to update AMI version on node groups?

### Actual Behavior

Updates to Node Group AMIs do not occur and stay stale as a result.

### Steps to Reproduce

1. `terraform apply` on existing EKS cluster that has node groups running on an older AMI version that the latest recommended.

**Comment by Dmitry1987** [comment]
from what I see in the repo, it's better to avoid using `aws_eks_node_group` and instead to use eksctl, because there's even no support for security group assignment at the moment in `aws_eks_node_group`. while eksctl is up to date with current aws features.

**Comment by congiv-VG** [comment]
> My interpretation of this was that we need to specify version to trigger Terraform drift detection that in turn would update release_version. https://github.com/hashicorp/terraform-provider-aws/issues/12675.

@gygitlab That's how I read it too. However, the only way I've gotten the behavior you're describing to work is by providing both `release_version` and `version`.

**Comment by mattburgess** [comment]
I'm not entirely convinced this is a bug, but am willing to be convinced otherwise :smile: 

What appears to be happening is:

1. At cluster *creation* time, with no `release_version` set in TF config, TF will ask EKS to create a node group without passing that parameter across in the API call.
2. EKS sees that API call and says "OK then, you didn't ask for a specific AMI version, I'll use the latest available today. Once created, it passes the AMI version it used back to Terraform
3. Terraform stores the AMI version in the `release_version` attribute in the state file
4. Some time passes, and you re-run `terraform plan`. As the `release_version` attribute is optional, and you've not specified it in the configuration, then Terraform won't (can't!) perform drift detection on it. 

So, in short, Terraform _will_ perform updates if either (or both) of `version` or `release_version` are specified in the config and change between runs.

**Comment by gygitlab** [comment]
> So, in short, Terraform will perform updates if **either** (or both) of version or release_version are specified in the config and change between runs.

So that's the point of the issue really. We _are_ specifying `version` here every time in our module and pinning it to the cluster version but it's not upgrading the node pool.

`release_version` is configurable by the user but if not it's set to `null`.

**Comment by mattburgess** [comment]
> `release_version` is configurable by the user but if not it's set to `null`.

That's not what I see here:

```hcl
resource "aws_eks_node_group" "t4g" {
  cluster_name    = aws_eks_cluster.example.name
  node_group_name = "example"
  node_role_arn   = aws_iam_role.eks_node.arn
  version         = aws_eks_cluster.example.version

  scaling_config {
    desired_size = 1
    min_size     = 1
    max_size     = 3
  }

  subnet_ids     = aws_subnet.eks_private.*.id
  instance_types = ["t4g.medium"]
  ami_type       = "AL2_ARM_64"

  lifecycle {
    ignore_changes = [scaling_config[0].desired_size]
  }

  timeouts {
    create = "5m"
  }

  depends_on = [
    aws_iam_role_policy_attachment.eks-node-AmazonEKSWorkerNodePolicy,
    aws_iam_role_policy_attachment.eks-node-AmazonEKS_CNI_Policy,
    aws_iam_role_policy_attachment.eks-node-AmazonEC2ContainerRegistryReadOnly,
    aws_iam_role_policy_attachment.eks-node-AmazonSSMManagedInstanceCore,
  ]
}
```

produces the following state (`terraform state show aws_eks_node_group.t4g`):
 
```hcl
resource "aws_eks_node_group" "t4g" {
    ami_type        = "AL2_ARM_64"
    arn             = "arn:aws:eks:eu-west-2:320797911953:nodegroup/example/example/f6c1d75a-71c7-b42a-dd95-77f60b71e950"
    capacity_type   = "ON_DEMAND"
    cluster_name    = "example"
    disk_size       = 20
    id              = "example:example"
    instance_types  = [
        "t4g.medium",
    ]
    labels          = {}
    node_group_name = "example"
    node_role_arn   = "arn:aws:iam::320797911953:role/eks-node"
    release_version = "1.22.12-20220926"
    resources       = [
        {
            autoscaling_groups              = [
                {
                    name = "eks-example-f6c1d75a-71c7-b42a-dd95-77f60b71e950"
                },
            ]
            remote_access_security_group_id = ""
        },
    ]
    status          = "ACTIVE"
    subnet_ids      = [
        "subnet-0542c30a2e2d244dd",
        "subnet-065b709446dfea43a",
        "subnet-09aed0a9d62486a4e",
    ]
    tags            = {}
    tags_all        = {}
    version         = "1.22"

    scaling_config {
        desired_size = 1
        max_size     = 3
        min_size     = 1
    }

    timeouts {
        create = "5m"
    }

    update_config {
        max_unavailable            = 1
        max_unavailable_percentage = 0
    }
}
```
Note specifically that, like you, my node group's `version` is pinned to the cluster's version and I don't specify a `release_version`, yet I get a `release_version` stored in my state.

Further, if I now apply an upgrade by bumping my cluster's version to `1.23` I get the following (expected) plan output:

```text
Terraform will perform the following actions:

  # aws_eks_cluster.example will be updated in-place
  ~ resource "aws_eks_cluster" "example" {
        id                        = "example"
        name                      = "example"
        tags                      = {}
      ~ version                   = "1.22" -> "1.23"
        # (10 unchanged attributes hidden)

        # (2 unchanged blocks hidden)
    }

  # aws_eks_node_group.t4g will be updated in-place
  ~ resource "aws_eks_node_group" "t4g" {
        id              = "example:example"
        tags            = {}
      ~ version         = "1.22" -> "1.23"
        # (14 unchanged attributes hidden)

        # (3 unchanged blocks hidden)
    }

Plan: 0 to add, 2 to change, 0 to destroy.
```

Resulting in the following state for my `aws_eks_node_group` resource (notably the `release_version` has been bumped):

```hcl
resource "aws_eks_node_group" "t4g" {
    ami_type        = "AL2_ARM_64"
    arn             = "arn:aws:eks:eu-west-2:320797911953:nodegroup/example/example/f6c1d75a-71c7-b42a-dd95-77f60b71e950"
    capacity_type   = "ON_DEMAND"
    cluster_name    = "example"
    disk_size       = 20
    id              = "example:example"
    instance_types  = [
        "t4g.medium",
    ]
    labels          = {}
    node_group_name = "example"
    node_role_arn   = "arn:aws:iam::320797911953:role/eks-node"
    release_version = "1.23.9-20220926"
    resources       = [
        {
            autoscaling_groups              = [
                {
                    name = "eks-example-f6c1d75a-71c7-b42a-dd95-77f60b71e950"
                },
            ]
            remote_access_security_group_id = ""
        },
    ]
    status          = "ACTIVE"
    subnet_ids      = [
        "subnet-0542c30a2e2d244dd",
        "subnet-065b709446dfea43a",
        "subnet-09aed0a9d62486a4e",
    ]
    tags            = {}
    tags_all        = {}
    version         = "1.23"

    scaling_config {
        desired_size = 1
        max_size     = 3
        min_size     = 1
    }

    timeouts {
        create = "5m"
    }

    update_config {
        max_unavailable            = 1
        max_unavailable_percentage = 0
    }
}
```

**Comment by gygitlab** [comment]
Ah I see @mattburgess. Apologies for the confusion on my part.

So where the confusion has arose here is that I'm talking about Node Group upgrades _when the Cluster hasn't been upgraded_. I.E. When AWS has released a new AMI for node groups on the same EKS Cluster Version.

I can see why this wouldn't work now though with the extra details here. There's two distinct fields and if neither are changing on Terraform's end then quite rightly it won't upgrade.

To that end then what it seems like we need is a data source similar to [`aws_eks_addon_version`](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/data-sources/eks_addon_version) where we can poll for the latest EKS Node Group AMI version and set that as `release_version` accordingly?

**Comment by mattburgess** [comment]
https://docs.aws.amazon.com/eks/latest/userguide/eks-optimized-ami.html contains a table of links for the various currently recommended AMIs. Those links take you to a AWS Systems Manager Parameter with the currently recommended AMI! So, this gets us some of the way there:

```hcl
data aws_ssm_parameter "eks_ami" {
    name = "/aws/service/eks/optimized-ami/${aws_eks_cluster.example.version}/amazon-linux-2/recommended/image_id"
}
```

Alas, here we run into a slight problem...neither the AMI name, nor any of its other attributes, seemingly has a 1:1 correlation with the `release_version` string we need. From my state file above, we can see we need to give the `aws_eks_node_group` a  `release_version = 1.23.9-20220926"`. But, the AMI's name only contains: `name = amazon-eks-node-1.23-v20220926`. If it contained `1.23.9`, at least we'd be able to do some admittedly potentially brittle substring parsing to get what we need out.

I think it might be worth pursuing with AWS whether or not it's feasible for them to either update their AMI naming to match the release_version scheme, or if they could at least put a tag/tags on the AMI with that info in it.

**Comment by mattburgess** [comment]
Doh, a bit more investigation and they do have what we need :smile:  Try this @gygitlab:

```hcl
data aws_ssm_parameter "eks_ami" {
    name = /aws/service/eks/optimized-ami/1.23/amazon-linux-2/recommended/release_version"
}
```

**Comment by gygitlab** [comment]
Thanks @mattburgess. I'll aim to try that soon.

**Comment by gygitlab** [comment]
Thanks again for this @mattburgess. So this does work but it's maybe not as ideal as a dedicated data source? Specifically the SSM param automatically gets marked as sensitive annoyingly due to limitations with that API I understand.

**Comment by mattburgess** [comment]
So I guess your issue is that even though the upgrade is triggered correctly, you don't know, specifically, what AMI release triggered it, i.e. what your configuration is going to end up in on apply?

I can see why you'd want that info up front but I wonder how gnarly the provider code would get if we tried to do a similar thing internally. We'd have to map the  `ami_type` to the correct SSM param and bail out if you've asked for an autoupgrade but are on a custom image. I might take a look at this over the next few days to see what it'd look like but can't promise anything.

**Comment by gygitlab** [comment]
No problem, thanks for taking a look.

**Comment by mattburgess** [comment]
Does the [`nonsensitive()`](https://www.terraform.io/language/functions/nonsensitive) function help here? I stumbled across it [here](https://github.com/hashicorp/terraform-provider-aws/issues/9090#issuecomment-818298866) whilst looking into coding up the workaround mentioned above. But I'm not sure that's needed anymore:

```hcl
resource "aws_eks_node_group" "t4g" {
  version         = aws_eks_cluster.example.version
  release_version = nonsensitive(data.aws_ssm_parameter.eks_ami_release_version.value)
  # other params
  ...
}
```

**Comment by gygitlab** [comment]
It does indeed work but that's a sniff I would argue. You don't really want to have to use that function as it could lead to confusion.

**Comment by dmcnaught** [comment]
@mattburgess Thanks for working on this.
We are using an `aws_launch_template` in our `aws_eks_node_group` and that doesn't take a `release_version`.
If I add `release_version` to the `aws_eks_node_group` in this configuration, no changes are detected.
Is there a way to do this using the `launch_template` option in the `aws_eks_node_group`.
Should I open a new ticket?
Thanks
--Duncan

**Comment by mattburgess** [comment]
@dmcnaught - the [API] docs(https://docs.aws.amazon.com/sdk-for-go/api/service/eks/#CreateNodegroupInput) say this:

> If you specify launchTemplate, and your launch template uses a custom AMI, then don't specify releaseVersion, or the node group deployment will fail. For more information about using launch templates with Amazon EKS, see Launch template support (https://docs.aws.amazon.com/eks/latest/userguide/launch-templates.html) in the Amazon EKS User Guide.

Now, if you're using a `launchTemplate` but are *not* using a custom AMI then I reckon you could achieve a similar thing by having your `aws_launch_template`'s `image_id` attribute track AWS' AMI releases. Something like this (completely untested!):

```hcl
data "aws_ssm_parameter" "eks_ami" {
    name = "/aws/service/eks/optimized-ami/${aws_eks_cluster.example.version}/amazon-linux-2/recommended/image_id"
}

resource "aws_launch_template" "eks_workers" {
    ...
    image_id = data.aws_ssm_parameter.eks_ami.value
    ...
}
```

In that scenario, you'd just not bother specifying `release_version` at all in your `eks_node_group` config.

**Comment by dmcnaught** [comment]
@mattburgess Thanks for the configuration.
It looks like there may be an issue with that since I got this error:
> Error: error updating EKS Node Group (...) version: InvalidParameterException: You cannot specify an image id within the launch template, since your nodegroup is configured to use an EKS optimized AMI.```

**Comment by mattburgess** [comment]
Ah, interesting! OK...what does `terraform state show aws_node_group.whatever` show for the `release_version` attribute? And does that match the latest version brought back by the `/aws/service/eks/optimized-ami/${aws_eks_cluster.example.version}/amazon-linux-2/recommended/release_version` SSM param? Given the behaviour I've observed, I can only imagine they're already on matching versions if you're not seeing any changes detected.

**Comment by dmcnaught** [comment]
@mattburgess 
Interesting - thanks. We upgraded the node group manually so that might have added the `release_version` to the node_group (which is there now and also on the latest - so it does match). Do you think it won't error when a new version is released?

**Comment by dmcnaught** [comment]
Hi @mattburgess ,
When we try to do the solution you outline, with
```hcl
data "aws_ssm_parameter" "eks_ami" {
    name = "/aws/service/eks/optimized-ami/${aws_eks_cluster.example.version}/amazon-linux-2/recommended/image_id"
}

resource "aws_launch_template" "eks_workers" {
    ...
    image_id = data.aws_ssm_parameter.eks_ami.value
    ...
}
```
The instances loose the Kubernetes bootstrap steps in the user data and so they don't connect to the cluster.
There is a note about that here:
https://github.com/terraform-aws-modules/terraform-aws-eks/blob/master/docs/user_data.md#summary
> If users supply an ami_id, the service no longers supplies user data to bootstrap nodes; users can enable enable_bootstrap_user_data and use the module provided user data template, or provide their own user data template

**Comment by mattburgess** [comment]
Interesting! That's confirmed by the upstream docs at https://docs.aws.amazon.com/eks/latest/userguide/launch-templates.html#launch-template-custom-ami (see the "Important" note at the bottom).

The bootstrap steps should therefore be explicitly passed in to the `aws_launch_template` using the `user_data` parameter of that resource. By the looks of that module's linux template, this looks to be the default user data, though there's another example in that doc I linked above:

```sh
#!/bin/bash
set -e
export SERVICE_IPV4_CIDR=${cluster_service_ipv4_cidr}
B64_CLUSTER_CA=${cluster_auth_base64}
API_SERVER_URL=${cluster_endpoint}
/etc/eks/bootstrap.sh ${cluster_name} ${bootstrap_extra_args} --b64-cluster-ca $B64_CLUSTER_CA --apiserver-endpoint $API_SERVER_URL
```
Obviously, you'll want to templatize that to pull in the relevant pieces from your cluster resource.
